File: telegram-bot-core/src/main/java/ru/tggc/telegrambotcore/access/checker/RoleAccessChecker.java
package ru.tggc.telegrambotcore.access.checker;

import com.pengrad.telegrambot.model.Chat;
import com.pengrad.telegrambot.model.User;
import com.pengrad.telegrambot.request.SendMessage;
import lombok.RequiredArgsConstructor;
import org.springframework.core.annotation.Order;
import org.springframework.stereotype.Component;
import ru.tggc.telegrambotcore.access.annotationprovider.AnnotationProviderFactory;
import ru.tggc.telegrambotcore.annotation.handle.HandleMeta;
import ru.tggc.telegrambotcore.dto.AccessResult;
import ru.tggc.telegrambotcore.dto.Response;
import ru.tggc.telegrambotcore.dto.UserRole;
import ru.tggc.telegrambotcore.service.UserService;

import java.lang.reflect.Method;

@Component
@Order(1)
@RequiredArgsConstructor
public class RoleAccessChecker implements AccessChecker {
    private final UserService userService;
    private final AnnotationProviderFactory factory;

    @Override
    public AccessResult check(User from, Method method, Chat chat) {
        HandleMeta annotationMeta = factory.getAnnotationMeta(method);
        UserRole[] roles = annotationMeta.requiredRoles();

        if (roles.length == 0) {
            return AccessResult.allow();
        }

        return userService.checkRoles(from.id(), roles)
                ? AccessResult.allow()
                : AccessResult.deny(Response.of(new SendMessage(chat.id().longValue(), "У вас недостаточно прав для это команды")));
    }
}

File: telegram-bot-core/src/main/kotlin/ru/tggc/telegrambotcore/access/checker/RoleAccessChecker.kt
package ru.tggc.telegrambotcore.access.checker

import com.pengrad.telegrambot.model.Chat
import com.pengrad.telegrambot.model.User
import com.pengrad.telegrambot.request.SendMessage
import org.springframework.core.annotation.Order
import org.springframework.stereotype.Component
import ru.tggc.telegrambotcore.access.annotationprovider.AnnotationProviderFactory
import ru.tggc.telegrambotcore.dto.AccessResult
import ru.tggc.telegrambotcore.dto.Response
import ru.tggc.telegrambotcore.service.UserService
import java.lang.reflect.Method

@Component
@Order(1)
class RoleAccessChecker(
    private val userService: UserService,
    private val factory: AnnotationProviderFactory
) : AccessChecker {
    override fun check(from: User, method: Method, chat: Chat): AccessResult {
        val roles = factory.getAnnotationMeta(method).requiredRoles

        if (roles.isEmpty() || userService.checkRoles(from.id(), roles)) {
            return AccessResult.allow()
        }

        return AccessResult.deny(
            Response.of {
                SendMessage(chat.id(), "У вас недостаточно прав для этой команды")
            }
        )
    }
}
